Fixed Point recipe (#196)
JoshOrndorff/recipes: text/3-entrees/fixed-point.md

# Fixed Point Arithmetic
*[`pallets/fixed-point`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/fixed-point)* *[`pallets/compounding-interest`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/compounding-interest)*

When programmers learn to use non-integer numbers in their programs, they are usually taught to use [floating point](https://en.wikipedia.org/wiki/Floating-point_arithmetic)s. In blockchain, we use an alternative representation of fractional numbers called [fixed point](https://en.wikipedia.org/wiki/Fixed-point_arithmetic). There are several ways to use fixed point numbers, and this recipe will introduce three of them. In particular we'll see:

* Substrate's own fixed point structs and traits
* The [substrate-fixed](https://github.com/encointer/substrate-fixed/) library
* A manual fixed point implementation (and why it's nicer to use a library)
* A comparison of the two libraries in a compounding interest example

## What's Wrong with Floats?

> Floats are cool for all kinds of reasons, but they also have one important drawback. Floating point arithmetic is **nondeterministic**. In order for the nodes in a blockchain network to reach agreement on the state of the chain, all operations must be completely deterministic. Luckily fixed point arithmetic is deterministic, and is often not much harder to use once you get the hang of it.

Reviewer: maybe you could also explain WHY floats are nondeterministic:  results can depend on hw, libs aso. There is an IEEE standard, but it's not 100% guaranteed for all implementations

Reviewer 2: We are planning to get a larger doc on DevHub in the next week or two about the problem with floating point and a summary of the tools available. Will definitely ask for you help there too if you don't mind. I think it would be better to dive deep there as Recipes tends to be less conceptual and jumps right into code. Like, "I already know that I want to do X, show me a pattern." 

Also happy to put more here as Joshy sees fit, just my two cents.

Author: I've added a tiny bit of details (only as much as @brenzi) mentioned in his comment. I'll leave the deep dive for the devhub knowledge base.

## Multiplicative Accumulators
*[`pallets/fixed-point`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/fixed-point)*

The first pallet covered in this recipe contains three implementations of a multiplicative accumulator. That's a fancy way to say the pallet lets users submit fractional numbers and keeps track of the product from multiplying them all together. The value starts out at one (the [multiplicative identity](https://en.wikipedia.org/wiki/Identity_element)), and it gets multiplied by whatever values the users submit. These three independent implementations compare and contrast the features of each.

### Permill Accumulator

> We'll using the most common approach which takes it's fixed point implementation from Substrate itself. There are a few fixed-point structs available in Substrate, all of which implement the [`PerThing` trait](https://substrate.dev/rustdocs/master/sp_arithmetic/trait.PerThing.html), that cover different amounts of precision. For this accumulator example, we'll use the [`PerMill` struct](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Permill.html) which represents fractions as parts per million. There are also [`Perbill`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Perbill.html), [`PerCent`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Percent.html), and [`PerU16`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.PerU16.html), which all provide the same interface (because it comes from the trait). Substrate's fixed-point structs are somewhat unique because they represent _only_ fractional parts of numbers. That means they can represent numbers between 0 and 1 inclusive, but _not_ numbers with whole parts like 2.718 or 3.14.

Reviewer 3: Possessive "its" has no apostrophe.

```suggestion
We'll using the most common approach which takes its fixed point implementation from Substrate itself. There are a few fixed-point structs available in Substrate, all of which implement the [`PerThing` trait](https://substrate.dev/rustdocs/master/sp_arithmetic/trait.PerThing.html), that cover different amounts of precision. For this accumulator example, we'll use the [`PerMill` struct](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Permill.html) which represents fractions as parts per million. There are also [`Perbill`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Perbill.html), [`PerCent`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Percent.html), and [`PerU16`](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.PerU16.html), which all provide the same interface (because it comes from the trait). Substrate's fixed-point structs are somewhat unique because they represent _only_ fractional parts of numbers. That means they can represent numbers between 0 and 1 inclusive, but _not_ numbers with whole parts like 2.718 or 3.14.
```

As in the `Permill` example, we begin by declaring the storage item. With substrate-fixed, there is not a `one` function, but there is a `from_num` function that we use to set the storage item's default value. This `from_num` method and it's counterpart `to-num` are your primary ways of converting between substrate-fixed types and Rust primitive types. If your use case does a lot of fixed-point arithmetic, like ours does, it is advisable to keep your data in substrate-fixed types.

> We're able to use `U16F16` as a storage item type because it, and the other substrate-fixed types, implement the parity scale codec.

```rust, ignore
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		// --snip--

		/// Substrate-fixed accumulator, value starts at 1 (multiplicative identity)

> 		FixedAccumulator get(fn fixed_value): U16F16 = U16F16::from_num(1);

Reviewer: I'm surprised that this works. Does this really set genesis value to 1?

Author: It doesn't set the genesis value, it sets the default value. That means when there is no value explicitly stored in that storage item the runtime will act as it 1 were stored.

There is a test for it here https://github.com/substrate-developer-hub/recipes/blob/joshy-fixed-point/pallets/fixed-point/src/tests.rs#L84

	}
}
```

Next we implement the extrinsic that allows users to update the accumulator by multiplying in a new value.

```rust, ignore
fn update_fixed(origin, new_factor: U16F16) -> DispatchResult {
	let _ = ensure_signed(origin)?;

	let old_accumulated = Self::fixed_value();

	// Multiply, handling overflow
	let new_product = old_accumulated.checked_mul(new_factor)
		.ok_or(Error::<T>::Overflow)?;

	// Write the new value to storage
	FixedAccumulator::put(new_product);

	// Emit event
	Self::deposit_event(Event::FixedUpdated(new_factor, new_product));
	Ok(())
}
```

> This extrinsic is quite similar to the `Permill` ersion with one notable difference. Because `U16F16` handles numbers greater than one, overflow is possible, and we need to handle it. The error handling here is straightforward, the important part is just that you remember to do it.

Reviewer: Your example is super simple and allows to compare the different approaches. However, if you don't mind, I'd mention the possibility to use additional math functions like *log*, *sqrt*, *exp*, *sin*, *pow*. Because that's what you can't do with the other options.

Author: I added a note mentioning that we'll see cooler features of substrate-fixed later on.

### Manual Accumulator

> In this final accumulator implementation, we manually track fixed point numbers using Rust's native `u32` as the underlying data type. This example is educational, but is only practical in the simplest scenarios. Generally you will have a more fun a less error-prone time coding if you use one of the previous two fixed-point types in your real-world applications.

Reviewer 3: ```suggestion
In this final accumulator implementation, we manually track fixed point numbers using Rust's native `u32` as the underlying data type. This example is educational, but is only practical in the simplest scenarios. Generally you will have a ~~more fun~~ less error-prone time coding if you use one of the previous two fixed-point types in your real-world applications.
```

Assuming the duplication was intentional levity.

```rust, ignore
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		// --snip--

		/// Balance for the discrete interest account
		DiscreteAccount get(fn discrete_account): u64;
	}
}
```

There are two extrinsics associated with the discrete interest account. The `deposit_discrete` extrinsic is shown here, and the `withdraw_discrete` extrinsic is nearly identical. Check it out in the kitchen.

```rust, ignore
fn deposit_discrete(origin, val_to_add: u64) -> DispatchResult {
	let _ = ensure_signed(origin)?;

	let old_value = DiscreteAccount::get();

	// Update storage for discrete account
	DiscreteAccount::put(old_value + val_to_add);

	// Emit event
	Self::deposit_event(Event::DepositedDiscrete(val_to_add));
	Ok(())
}
```

The flow of these deposit and withdraw extrinsics is entirely straight-forward. They each perform a simple addition or substraction from the stored value, and they have nothing to do with interest.

Because the interest is paid discretely every ten blocks it can be handled independently of deposits and withdrawals. The interest calculation happens automatically in the `on_finalize` block.

```rust, ignore

> fn on_finalize(n: T::BlockNumber) {

Reviewer: this function has O(n) complexity, n being the number of accounts. For the sake of simplicity, I understand why you do it like this and you only feature one account here, but it's not a best practise example.
I'd rather calculate compound interest on the spot when a balance is queried or updated. (i.e.: I will soon implement a *balances* module that features demurrage.... perishable money ;-). I'll let you know when it is ready.

Author: Yeah, all the caveats are true. And as you noticed I do it the way you describe in the substrate-fixed implementation. I think I'll leave it like this for the sake of comparison and contrast.

	if (n % 10.into()).is_zero() {

		// Calculate interest Interest = principle * rate * time
		// We can use the `*` operator for multiplying a `Percent` by a u64
		// because `Percent` implements the trait Mul<u64>
		let interest = Self::discrete_interest_rate() * DiscreteAccount::get() * 10;

		// The following line, although similar, does not work because
		// u64 does not implement the trait Mul<Percent>
		// let interest = DiscreteAccount::get() * Self::discrete_interest_rate() * 10;

		// Update the balance
		let old_balance = DiscreteAccount::get();
		DiscreteAccount::put(old_balance + interest);

		// Emit the event
		Self::deposit_event(Event::DiscreteInterestApplied(interest));
	}
}
```

`on_finalize` is called at the end of every block, but we only want to pay interest every ten blocks, so the first thing we do is check whether this block is a multiple of ten. If it is we calculate the interest due by the formula `interest = principle * rate * time`. As the comments explain, there is some subtlety in the order of the multiplication. You can multiply `PerCent * u64` but not `u64 * PerCent`.

### Continuously Compounding

> You can imagine increasing the frequency a which the interest is paid out. Increasing the frequency enough approaches [continuously compounding interest](https://en.wikipedia.org/wiki/Compound_interest#Continuous_compounding). Calculating continuously compounding interest requires the [exponential function](https://en.wikipedia.org/wiki/Exponential_function) which is not available using Substrate's `PerThing` types. Luckily exponential and other [transcendental functions](https://en.wikipedia.org/wiki/Transcendental_function) are available in substrate-fixed, which is why we've chosen to use it for this example.

Reviewer: ah, ok, forget my above comment ;-)

Reviewer: I should read first and write then ;-)

> You can imagine increasing the frequency a which the interest is paid out. Increasing the frequency enough approaches [continuously compounding interest](https://en.wikipedia.org/wiki/Compound_interest#Continuous_compounding). Calculating continuously compounding interest requires the [exponential function](https://en.wikipedia.org/wiki/Exponential_function) which is not available using Substrate's `PerThing` types. Luckily exponential and other [transcendental functions](https://en.wikipedia.org/wiki/Transcendental_function) are available in substrate-fixed, which is why we've chosen to use it for this example.

Reviewer 3: ```suggestion
You can imagine increasing the frequency at which the interest is paid out. Increasing the frequency enough approaches [continuously compounding interest](https://en.wikipedia.org/wiki/Compound_interest#Continuous_compounding). Calculating continuously compounding interest requires the [exponential function](https://en.wikipedia.org/wiki/Exponential_function) which is not available using Substrate's `PerThing` types. Luckily exponential and other [transcendental functions](https://en.wikipedia.org/wiki/Transcendental_function) are available in substrate-fixed, which is why we've chosen to use it for this example.
```

> With continuously compounded interest, we _could_ update the interest in `on_finalize` as we did before, but it woudl need to be updated every single block. Instead we wait until a user tries to use the account (to deposit or withdraw funds), and then calculate the account's current value "just in time".

Reviewer 3: ```suggestion
With continuously compounded interest, we _could_ update the interest in `on_finalize` as we did before, but it would need to be updated every single block. Instead we wait until a user tries to use the account (to deposit or withdraw funds), and then calculate the account's current value "just in time".
```

	}
}
```

As before, there are two relevant extrinsics, `deposit_continuous` and `withdraw_continuous`. They are nearly identical so we'll only show one.

```rust, ignore
fn deposit_continuous(origin, val_to_add: u64) -> DispatchResult {
	let _ = ensure_signed(origin)?;

	let current_block = system::Module::<T>::block_number();
	let old_value = Self::value_of_continuous_account(&current_block);

	// Update storage for compounding account
	ContinuousAccount::<T>::put(
		ContinuousAccountData {
			principle: old_value + I32F32::from_num(val_to_add),
			deposit_date: current_block,
		}
	);

	// Emit event
	Self::deposit_event(Event::DepositedContinuous(val_to_add));
	Ok(())
}
```

This function itself isn't too insightful. It does the same basic things as the discrete variant: lookup the old value, and the deposit, update storage, emit event. The one interesting part is that it calls a helper function to get the account's previous value. This helper function calculates the value of the account considering all the interest that has accrued since the last time the account was touched. Let's take a closer look.

```rust, ignore

> fn value_of_continuous_account(now: &<T as system::Trait>::BlockNumber) -> I32F32 {

Reviewer: nice, that's my demurrage example in reverse....

Author: I look forward to seeing the example!

	// Get the old state of the accout
	let ContinuousAccountData{
		principle,
		deposit_date,
	} = ContinuousAccount::<T>::get();

	// Calculate the exponential function (lots of type conversion)
	let elapsed_time_block_number = *now - deposit_date;
	let elapsed_time_u32 = TryInto::try_into(elapsed_time_block_number).ok()

> 		.expect("blockchain will not exceed 2^32 blocks; qed");

Reviewer 3: ```suggestion
	let elapsed_time_u32 = TryInto::try_into(elapsed_time_block_number)
		.expect("blockchain will not exceed 2^32 blocks; qed");
```

`.expect()` exists on both `Option` and `Result`.

# Fixed Point Arithmetic

> *[`pallets/fixed-point`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/fixed-point)* *[`pallets/compounding-interest`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/compounding-interest)*

Reviewer 2: Is this line normally here?

Author: Yes. I've been placing a link from the text to the code it describes in each write-up. Here are a few [other](https://github.com/substrate-developer-hub/recipes/blame/master/text/3-entrees/basic-pow.md#L2) [examples](https://github.com/substrate-developer-hub/recipes/blame/master/text/3-entrees/constants.md#L2).

This one is slightly unique because this write-up is describing two pallets rather than one.

The code of this extrinsic largely speaks for itself. One thing to take particular note of is that we _don't_ check for overflow on the multiplication. If you've read many of the recipes you know that a Substrate runtime must never panic, and a developer must be extremely diligent in always checking for and gracefully handling error conditions. Because `Permill` only holds values between 0 and 1, we know that their product will always be in that same range. Thus it is impossible to overflow or saturate. So we can happily use `saturating_mul` and move on.

### Substrate-fixed Accumulator

[Substrate-fixed](https://github.com/encointer/substrate-fixed/) takes a more traditional approach in that their types represent numbers with both whole _and_ fractional parts. For this implementation, we'll use the `U16F16` type. This type contains an unsigned number (indicated by the `U` at the beginning) and has 32 _total_ bits of precision - 16 for the integer part, and 16 for the fractional part. There are several other types provided that follow the same naming convention. Some examples include `U32F32` and `I32F32` where the `I` indicates a signed number, just like in Rust primitive types.

> As in the `Permill` example, we begin by declaring the storage item. With substrate-fixed, there is not a `one` function, but there is a `from_num` function that we use to set the storage item's default value. This `from_num` method and it's counterpart `to-num` are your primary ways of converting between substrate-fixed types and Rust primitive types. If your use case does a lot of fixed-point arithmetic, like ours does, it is advisable to keep your data in substrate-fixed types.

Reviewer 2: ```suggestion
As in the `Permill` example, we begin by declaring the storage item. With substrate-fixed, there is not a `one` function, but there is a `from_num` function that we use to set the storage item's default value. This `from_num` method and its counterpart `to-num` are your primary ways of converting between substrate-fixed types and Rust primitive types. If your use case does a lot of fixed-point arithmetic, like ours does, it is advisable to keep your data in substrate-fixed types.
```

> > We're able to use `U16F16` as a storage item type because it, and the other substrate-fixed types, implement the parity scale codec.

Reviewer 2: ```suggestion
> We're able to use `U16F16` as a storage item type because it, and the other substrate-fixed types, implements the parity scale codec.
```

	ensure_signed(origin)?;

	let old_accumulated = Self::fixed_value();

	// Multiply, handling overflow
	let new_product = old_accumulated.checked_mul(new_factor)
		.ok_or(Error::<T>::Overflow)?;

	// Write the new value to storage
	FixedAccumulator::put(new_product);

	// Emit event
	Self::deposit_event(Event::FixedUpdated(new_factor, new_product));
	Ok(())
}
```

This extrinsic is quite similar to the `Permill` version with one notable difference. Because `U16F16` handles numbers greater than one, overflow is possible, and we need to handle it. The error handling here is straightforward, the important part is just that you remember to do it.

This example has shown the fundamentals of substrate-fixed, but this library has much more to offer as we'll see in the compounding interest example.

### Manual Accumulator

In this final accumulator implementation, we manually track fixed point numbers using Rust's native `u32` as the underlying data type. This example is educational, but is only practical in the simplest scenarios. Generally you will have a ~~more fun~~ less error-prone time coding if you use one of the previous two fixed-point types in your real-world applications.

> Fixed point is not very complex conceptually. We represent fractional numbers as regular old integers, and we decide in advance to consider some of the place values fractional. It's just like saying we'll omit the decimal point when talking about money and all agree that "1995" actually _means_ €19.95. This is exactly how Substrate's [Balances pallet](https://substrate.dev/rustdocs/master/pallet_balances/index.html) works, a tradition that's been in blockchain since Bitcon. In our example we will treat 16 bits as integer values, and 16 as fractional, just as substrate-fixed's `U16F16` did.

Reviewer 2: Not 100% sure on this, but I think only US puts the $ before the number. 
```suggestion
Fixed point is not very complex conceptually. We represent fractional numbers as regular old integers, and we decide in advance to consider some of the place values fractional. It's just like saying we'll omit the decimal point when talking about money and all agree that "1995" actually _means_ 19.95 €. This is exactly how Substrate's [Balances pallet](https://substrate.dev/rustdocs/master/pallet_balances/index.html) works, a tradition that's been in blockchain since Bitcon. In our example we will treat 16 bits as integer values, and 16 as fractional, just as substrate-fixed's `U16F16` did.
```

	Ok(())
}
```

As mentioned above, when you multiply two 32-bit numbers, you can end up with as much as 64 bits in the product. In this 64-bit intermediate product, we have 32 integer bits and 32 fractional. We can simply throw away the 16 right-most fractional bits that merely provide extra precision. But we need to be careful with the 16 left-most integer bits. If any of those bits are non-zero after the multiplication it means overflow has occurred. If they are all zero, then we can safely throw them away as well.

> If this business about having more bits after the multiplication is confusing, try this exercise in the more familiar decimal system. Consider these numbers that have 4 total digits (2 integer, and two fractional): 12.34 and 56.78. Multiply them together. How many integer and fractional digits are in the product? Try that again with larger numbers like 01.23 * 02.48 and smaller like 11.11 and 22.22. Which of these products can be fit back into a 4-digit number like the ones we started with?

## Compounding Interest
*[`pallets/compounding-interest`](https://github.com/substrate-developer-hub/recipes/tree/master/pallets/compounding-interest)*

> Many financial agreements involve interest for loaned or borrowed money. [Compounding interest](https://en.wikipedia.org/wiki/Compound_interest) is when new interest is paid on top of not only the original loan amount, the so-called "principle", but also any interest that has been previously paid.

Reviewer 2: Only the "moral value/code" definition uses the "principle" spelling, all other definitions are "principal".
```suggestion
Many financial agreements involve interest for loaned or borrowed money. [Compounding interest](https://en.wikipedia.org/wiki/Compound_interest) is when new interest is paid on top of not only the original loan amount, the so-called "principal", but also any interest that has been previously paid.
```

Reviewer 2: Should also change in the code

Reviewer 2: Should also fix in the code

### Discrete Compounding

Our first example will look at discrete compounding interest. This is when interest is paid at a fixed interval. In our case, interest will be paid every ten blocks.

> For this implementation we've chosen to use Substrate's [`Percent` type](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Percent.html). It works nearly the same as `Permill`, but it represents numbers as "parts per hundred" rather than "parts er million". We could also have used Substrate-fixed for this implementation, but chose to save it for the next example.

Reviewer 2: ```suggestion
For this implementation we've chosen to use Substrate's [`Percent` type](https://substrate.dev/rustdocs/master/sp_arithmetic/struct.Percent.html). It works nearly the same as `Permill`, but it represents numbers as "parts per hundred" rather than "parts per million". We could also have used Substrate-fixed for this implementation, but chose to save it for the next example.
```

Because the interest is paid discretely every ten blocks it can be handled independently of deposits and withdrawals. The interest calculation happens automatically in the `on_finalize` block.

```rust, ignore
fn on_finalize(n: T::BlockNumber) {
	// Apply newly-accrued discrete interest every ten blocks
	if (n % 10.into()).is_zero() {

		// Calculate interest Interest = principle * rate * time
		// We can use the `*` operator for multiplying a `Percent` by a u64
		// because `Percent` implements the trait Mul<u64>
		let interest = Self::discrete_interest_rate() * DiscreteAccount::get() * 10;

		// The following line, although similar, does not work because
		// u64 does not implement the trait Mul<Percent>
		// let interest = DiscreteAccount::get() * Self::discrete_interest_rate() * 10;

		// Update the balance
		let old_balance = DiscreteAccount::get();
		DiscreteAccount::put(old_balance + interest);

		// Emit the event
		Self::deposit_event(Event::DiscreteInterestApplied(interest));
	}
}
```

> `on_finalize` is called at the end of every block, but we only want to pay interest every ten blocks, so the first thing we do is check whether this block is a multiple of ten. If it is we calculate the interest due by the formula `interest = principle * rate * time`. As the comments explain, there is some subtlety in the order of the multiplication. You can multiply `PerCent * u64` but not `u64 * PerCent`.

Reviewer 2: ```suggestion
`on_finalize` is called at the end of every block, but we only want to pay interest every ten blocks, so the first thing we do is check whether this block is a multiple of ten. If it is we calculate the interest due by the formula `interest = principal * rate * time`. As the comments explain, there is some subtlety in the order of the multiplication. You can multiply `PerCent * u64` but not `u64 * PerCent`.
```

With the struct to represent the account's state defined, we can initialize the storage value.

```rust, ignore
decl_storage! {
	trait Store for Module<T: Trait> as Example {
		// --snip--

		/// Balance for the continuously compounded account
		ContinuousAccount get(fn balance_compound): ContinuousAccountData<T::BlockNumber>;
	}
}
```

As before, there are two relevant extrinsics, `deposit_continuous` and `withdraw_continuous`. They are nearly identical so we'll only show one.

```rust, ignore
fn deposit_continuous(origin, val_to_add: u64) -> DispatchResult {
	ensure_signed(origin)?;

	let current_block = system::Module::<T>::block_number();
	let old_value = Self::value_of_continuous_account(&current_block);

	// Update storage for compounding account
	ContinuousAccount::<T>::put(
		ContinuousAccountData {
			principle: old_value + I32F32::from_num(val_to_add),
			deposit_date: current_block,
		}
	);

	// Emit event
	Self::deposit_event(Event::DepositedContinuous(val_to_add));
	Ok(())
}
```

> This function itself isn't too insightful. It does the same basic things as the discrete variant: lookup the old value, and the deposit, update storage, emit event. The one interesting part is that it calls a helper function to get the account's previous value. This helper function calculates the value of the account considering all the interest that has accrued since the last time the account was touched. Let's take a closer look.

Reviewer 2: ```suggestion
This function itself isn't too insightful. It does the same basic things as the discrete variant: look up the old value and the deposit, update storage, and emit an event. The one interesting part is that it calls a helper function to get the account's previous value. This helper function calculates the value of the account considering all the interest that has accrued since the last time the account was touched. Let's take a closer look.
```
